# *Project: Vanguard A/B Test Results Analysis*
---

### CONTEXT:
- Company : Vanguard, the US-based investment management company. (website: https://investor.vanguard.com/)- Role : newly hired DATA ANALYST in the Customer Experience (CX) team. 
The team launched an exciting digital experiment, and now, they're eagerly waiting to uncover the results and need your help!
- Task : Analyze the results of the digital experiment conducted by the team.Primary objective:  Decode the experiment's performance.
- The critical question : Would these changes encourage more clients to complete the process?
- Belief : Vanguard believed that a more intuitive, modern UI with timely in-context prompts (cues, messages, or instructions within the user’s current task) could make the online process smoother for clients.
---
An **A/B test** was set **into motion** `from 3/15/2017 to 6/20/2017` by the team.
- Control Group: Clients interacted with Vanguard's traditional online process. (old UI)- Test Group: Clients experienced the new, spruced-up digital interface. (new UI)
---

Both groups navigated through an identical process sequence:
- an initial page (start), 
- three subsequent steps (step 1, step 2, step 3), 
- and finally, a confirmation page signaling process completion.
---
**The goal is to see if the new design leads to a better user experience and higher process completion rates.**

---

*Datasets and CSV files*

| Dataset Name                        | DataFrame Name                                               | Description                                                                 |
|---------------------------------|--------------------------------------------------------------|-----------------------------------------------------------------------------|
| **Client Profiles**              | `df_final_demo`                                             | Demographics of clients including age, gender, and account details.        |
| **Digital Footprints (Part 1 & 2)** | `df_digital_footprints` (merge of `df_pt_1` and `df_pt_2`) | Detailed trace of client online interactions; parts 1 & 2 should be merged.|
| **Experiment Roster**            | `df_final_experiment_clients`                                | List of clients who participated in the grand experiment.                  |


---
## *Dataset 1: CLIENTS PROFILES*
- CSV: `../data/interim/client_profiles.csv`
- DataFrame: `df_final_demo`
- **Demographics like age, gender, and account details of our clients.**

In [ ]:
import pandas as pd
from IPython.display import display

# Load the df_final_demo text file and convert to DataFrame then save to CSV ---------------------------------------------------------------------------------------------

# Load df_final_demo dataset and split by the separator comma ","
df_final_demo = pd.read_csv("../data/raw/df_final_demo.txt",sep=",")

# Save the DataFrame to CSV ---------------------------------------------------------------------------------------------
print(f"DEMO DATA shape: {df_final_demo.shape}")
df_final_demo.to_csv("../data/interim/client_profiles.csv",index=False)

# Display the DataFrame to ensure it loaded correctly --------------------------------------------------------------
print("Saved df_final_demo in a CSV file located ../data/interim/client_profiles.csv.")
display(df_final_demo)

# Assign df_final_demo to df_clients_profiles
df_clients_profiles = df_final_demo


DEMO DATA shape: (70609, 9)
Saved df_final_demo in a CSV file located ../data/interim/client_profiles.csv.


client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
70604    7993686             4.0              56.0      38.5     U        3.0   
70605    8981690            12.0             148.0      31.0     M        2.0   
70606     333913            16.0             198.0      61.5     F        2.0   
70607    1573142            21.0             255.0      68.0     M        3.0   
70608    5602139            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth  
0        45105.30           6.0            9.0  
1       110860.30           6.0            9.0  
2        52467.79           6.0            9.0  
3        67454.65           3.0            6.0  
4       103671.75           0.0            3.0  
...           ...           ...            ...  
70604  1411062.68           5.0            5.0  
70605   101867.07           6.0            6.0  
70606    40745.00           3.0            3.0  
70607   475114.69           4.0            4.0  
70608   157498.73           7.0            7.0  

[70609 rows x 9 columns]

---
## *Dataset 2: DIGITAL FOOTPRINTS*
- CSV: `../data/interim/df_digital_footprints.csv`
- DataFrame: `df_digital_footprints`
- Merge of `pt_1` and `pt_2`

---
### Merging **pt_1** and **pt_2** into final dataset for **df_digital_footprints**
FROM TXT:
- `pt_1.txt`
- `pt_2.txt`

TO CSV:
- `digital_footprints_pt_1.csv`
- `digital_footprints_pt_2.csv`
- `digital_footprints_final.csv`

In [ ]:
import pandas as pd
from IPython.display import display
# Load the txt files (comma-separated) and convert to DataFrames then save to CSV ---------------------------------------------------------------------------------------------
# Load the two parts of the final web data and split by the separator comma ","
df_pt_1 = pd.read_csv("../data/raw/df_final_web_data_pt_1.txt",sep=",")
df_pt_2 = pd.read_csv("../data/raw/df_final_web_data_pt_2.txt",sep=",")

# Display the 2 DataFrames to ensure they loaded correctly --------------------------------------------------------------
display(df_pt_1.head())
display(df_pt_2.head())

# Save each DataFrame to CSV ---------------------------------------------------------------------------------------------
print(f"WEB DATA PART 1 shape: {df_pt_1.shape}")
df_pt_1.to_csv("../data/interim/digital_footprints_pt_1.csv",index=False)
print("Saved df_pt_1 in ../data/interim/digital_footprints_pt_1.csv")

print(f"WEB DATA PART 2 shape: {df_pt_2.shape}")
df_pt_2.to_csv("../data/interim/digital_footprints_pt_2.csv",index=False)
print("Saved df_pt_2 in ../data/interim/digital_footprints_pt_2.csv")
print("Saved both parts as CSV files.")

# Merge the two parts (same columns) --------------------------------------------------------------------------------
df_final_web_data = pd.concat([df_pt_1, df_pt_2],ignore_index=True)
print(f"Final Merged part shape: {df_final_web_data.shape}")

# Save merged CSV ---------------------------------------------------------------------------------------------
df_final_web_data.to_csv('../data/interim/digital_footprints.csv', index=False)

print("Merged digital footprints parts and saved in '../data/interim/digital_footprints.csv'")

# Display the merged DataFrame to ensure it loaded correctly ---------------------------------------------------------------------------------------------
display(df_final_web_data.head())

# Assign merged df to df_digital_footprints------------------------------------------------------------------
df_digital_footprints = df_final_web_data

client_id            visitor_id                      visit_id process_step  \
0    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
1    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
2    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
3    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
4    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   

             date_time  
0  2017-04-17 15:27:07  
1  2017-04-17 15:26:51  
2  2017-04-17 15:19:22  
3  2017-04-17 15:19:13  
4  2017-04-17 15:18:04

client_id             visitor_id                      visit_id  \
0     763412  601952081_10457207388  397475557_40440946728_419634   
1    6019349  442094451_91531546617  154620534_35331068705_522317   
2    6019349  442094451_91531546617  154620534_35331068705_522317   
3    6019349  442094451_91531546617  154620534_35331068705_522317   
4    6019349  442094451_91531546617  154620534_35331068705_522317   

  process_step            date_time  
0      confirm  2017-06-06 08:56:00  
1      confirm  2017-06-01 11:59:27  
2       step_3  2017-06-01 11:58:48  
3       step_2  2017-06-01 11:58:08  
4       step_1  2017-06-01 11:57:58

WEB DATA PART 1 shape: (343141, 5)
Saved pt_1 in ../data/interim/digital_footprints_pt_1.csv
WEB DATA PART 2 shape: (412264, 5)
Saved pt_2 in ../data/interim/digital_footprints_pt_2.csv
Saved both parts as CSV files.
Final Merged part shape: (755405, 5)
Merged digital footprints parts and saved in '../data/interim/digital_footprints.csv'


client_id            visitor_id                      visit_id process_step  \
0    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
1    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
2    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
3    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
4    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   

             date_time  
0  2017-04-17 15:27:07  
1  2017-04-17 15:26:51  
2  2017-04-17 15:19:22  
3  2017-04-17 15:19:13  
4  2017-04-17 15:18:04

In [4]:
import pandas as pd
# load csv to verify
df_digital_footprints = pd.read_csv('../data/interim/digital_footprints.csv')
display(df_digital_footprints)

client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
755400    9668240   388766751_9038881013   922267647_3096648104_968866   
755401    9668240   388766751_9038881013   922267647_3096648104_968866   
755402    9668240   388766751_9038881013   922267647_3096648104_968866   
755403    9668240   388766751_9038881013   922267647_3096648104_968866   
755404     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
755400        start  2017-05-24 18:46:10  
755401        start  2017-05-24 18:45:29  
755402       step_1  2017-05-24 18:44:51  
755403        start  2017-05-24 18:44:34  
755404        start  2017-06-03 12:17:09  

[755405 rows x 5 columns]

---
## *Dataset 3: EXPERIMENT ROSTER*
- CSV: `../data/interim/experiment_roster.csv`
- DataFrame: `df_final_experiment_clients`
- A list revealing which clients were part of the grand experiment

In [5]:
import pandas as pd
from IPython.display import display

# Load the df_final_experiment_clients. text file and convert to DataFrame then save to CSV ---------------------------------------------------------------------------------------------

# Load df_final_demo dataset and split by the separator comma ","
df_final_experiment_clients = pd.read_csv("../data/raw/df_final_experiment_clients.txt",sep=",")

# Save the DataFrame to CSV ---------------------------------------------------------------------------------------------
print(f"FINAL EXPERIMENT CLIENTS shape: {df_final_experiment_clients.shape}")
df_final_experiment_clients.to_csv("../data/interim/experiment_roster.csv",index=False)

# Display the DataFrame to ensure it loaded correctly --------------------------------------------------------------
print("Saved df_final_experiment_clients as CSV file in ../data/interim/experiment_roster.csv.")
display(df_final_experiment_clients)

# Assign df_final_experiment_clients to df_experiment_roster
df_experiment_roster = df_final_experiment_clients

FINAL EXPERIMENT CLIENTS shape: (70609, 2)
Saved df_final_experiment_clients as CSV file in ../data/interim/experiment_roster.csv.


client_id Variation
0        9988021      Test
1        8320017      Test
2        4033851   Control
3        1982004      Test
4        9294070   Control
...          ...       ...
70604    2443347       NaN
70605    8788427       NaN
70606     266828       NaN
70607    1266421       NaN
70608    9895983       NaN

[70609 rows x 2 columns]

---
## *Load the 3 DATASETS into DataFrames*
- `df_client_profiles` loaded with **"../data/interim/client_profiles.csv"** 
- `df_digital_footprints` loaded with  **"../data/interim/digital_footprints.csv"**
- `df_experiment_roster` loaded with **"../data/interim/experiment_roster.csv"**

In [2]:
import pandas as pd

# client_profiles dataset
df_client_profiles = pd.read_csv("../data/interim/client_profiles.csv")

# digital_footprints dataset (main dataset for client behaviour)
df_digital_footprints = pd.read_csv("../data/interim/digital_footprints.csv")

# experiment_roster dataset
df_experiment_roster = pd.read_csv("../data/interim/experiment_roster.csv")


---
## *Explore the 3 DATASETS seperately*

In [7]:
import pandas as pd
# display the datasets ---------------------------------------------------------------------------------------------

# display client_profiles dataset
print("Client Profiles Dataset:")
print("Client Profiles Dataset Shape:", df_client_profiles.shape[0], "rows and", df_client_profiles.shape[1], "columns\n")
df_client_profiles.info()
display(df_client_profiles)

# display digital_footprints dataset
print("Digital Footprints Dataset:")
print("Digital Footprints Dataset Shape:", df_digital_footprints.shape[0], "rows and", df_digital_footprints.shape[1], "columns\n")
df_digital_footprints.info()
display(df_digital_footprints)

# display experiment_roster dataset
print("Experiment Roster Dataset:")
print("Experiment Roster Dataset Shape:", df_experiment_roster.shape[0], "rows and", df_experiment_roster.shape[1], "columns\n")
df_experiment_roster.info()
display(df_experiment_roster)


Client Profiles Dataset:
Client Profiles Dataset Shape: 70609 rows and 9 columns

<class 'pandas.DataFrame'>
RangeIndex: 70609 entries, 0 to 70608
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   client_id         70609 non-null  int64  
 1   clnt_tenure_yr    70595 non-null  float64
 2   clnt_tenure_mnth  70595 non-null  float64
 3   clnt_age          70594 non-null  float64
 4   gendr             70595 non-null  str    
 5   num_accts         70595 non-null  float64
 6   bal               70595 non-null  float64
 7   calls_6_mnth      70595 non-null  float64
 8   logons_6_mnth     70595 non-null  float64
dtypes: float64(7), int64(1), str(1)
memory usage: 4.8 MB


client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0         836976             6.0              73.0      60.5     U        2.0   
1        2304905             7.0              94.0      58.0     U        2.0   
2        1439522             5.0              64.0      32.0     U        2.0   
3        1562045            16.0             198.0      49.0     M        2.0   
4        5126305            12.0             145.0      33.0     F        2.0   
...          ...             ...               ...       ...   ...        ...   
70604    7993686             4.0              56.0      38.5     U        3.0   
70605    8981690            12.0             148.0      31.0     M        2.0   
70606     333913            16.0             198.0      61.5     F        2.0   
70607    1573142            21.0             255.0      68.0     M        3.0   
70608    5602139            21.0             254.0      59.5     F        3.0   

              bal  calls_6_mnth  logons_6_mnth  
0        45105.30           6.0            9.0  
1       110860.30           6.0            9.0  
2        52467.79           6.0            9.0  
3        67454.65           3.0            6.0  
4       103671.75           0.0            3.0  
...           ...           ...            ...  
70604  1411062.68           5.0            5.0  
70605   101867.07           6.0            6.0  
70606    40745.00           3.0            3.0  
70607   475114.69           4.0            4.0  
70608   157498.73           7.0            7.0  

[70609 rows x 9 columns]

Digital Footprints Dataset:
Digital Footprints Dataset Shape: 755405 rows and 5 columns

<class 'pandas.DataFrame'>
RangeIndex: 755405 entries, 0 to 755404
Data columns (total 5 columns):
 #   Column        Non-Null Count   Dtype
---  ------        --------------   -----
 0   client_id     755405 non-null  int64
 1   visitor_id    755405 non-null  str  
 2   visit_id      755405 non-null  str  
 3   process_step  755405 non-null  str  
 4   date_time     755405 non-null  str  
dtypes: int64(1), str(4)
memory usage: 28.8 MB


client_id             visitor_id                      visit_id  \
0         9988021   580560515_7732621733  781255054_21935453173_531117   
1         9988021   580560515_7732621733  781255054_21935453173_531117   
2         9988021   580560515_7732621733  781255054_21935453173_531117   
3         9988021   580560515_7732621733  781255054_21935453173_531117   
4         9988021   580560515_7732621733  781255054_21935453173_531117   
...           ...                    ...                           ...   
755400    9668240   388766751_9038881013   922267647_3096648104_968866   
755401    9668240   388766751_9038881013   922267647_3096648104_968866   
755402    9668240   388766751_9038881013   922267647_3096648104_968866   
755403    9668240   388766751_9038881013   922267647_3096648104_968866   
755404     674799  947159805_81558194550   86152093_47511127657_716022   

       process_step            date_time  
0            step_3  2017-04-17 15:27:07  
1            step_2  2017-04-17 15:26:51  
2            step_3  2017-04-17 15:19:22  
3            step_2  2017-04-17 15:19:13  
4            step_3  2017-04-17 15:18:04  
...             ...                  ...  
755400        start  2017-05-24 18:46:10  
755401        start  2017-05-24 18:45:29  
755402       step_1  2017-05-24 18:44:51  
755403        start  2017-05-24 18:44:34  
755404        start  2017-06-03 12:17:09  

[755405 rows x 5 columns]

Experiment Roster Dataset:
Experiment Roster Dataset Shape: 70609 rows and 2 columns

<class 'pandas.DataFrame'>
RangeIndex: 70609 entries, 0 to 70608
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   client_id  70609 non-null  int64
 1   Variation  50500 non-null  str  
dtypes: int64(1), str(1)
memory usage: 1.1 MB


client_id Variation
0        9988021      Test
1        8320017      Test
2        4033851   Control
3        1982004      Test
4        9294070   Control
...          ...       ...
70604    2443347       NaN
70605    8788427       NaN
70606     266828       NaN
70607    1266421       NaN
70608    9895983       NaN

[70609 rows x 2 columns]

---
## *Combine the 3 datasets into 1 dataset*
- The 3 datasets share client_id as a common column

In [ ]:
import pandas as pd

# Load datasets
df_client_profiles = pd.read_csv("../data/interim/client_profiles.csv")
df_digital_footprints = pd.read_csv("../data/interim/digital_footprints.csv")
df_experiment_roster = pd.read_csv("../data/interim/experiment_roster.csv")

# Merge datasets on client_id (left joins)
df_ab_test = df_client_profiles.merge(df_digital_footprints, on="client_id", how="left") \
                            .merge(df_experiment_roster, on="client_id", how="left")

# Save as main dataset
df_ab_test.to_csv("../data/interim/vanguard_ab_test.csv", index=False)


In [ ]:
display(df_ab_test.head())

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0     836976             6.0              73.0      60.5     U        2.0   
1     836976             6.0              73.0      60.5     U        2.0   
2     836976             6.0              73.0      60.5     U        2.0   
3     836976             6.0              73.0      60.5     U        2.0   
4     836976             6.0              73.0      60.5     U        2.0   

       bal  calls_6_mnth  logons_6_mnth            visitor_id  \
0  45105.3           6.0            9.0  427070339_1413275162   
1  45105.3           6.0            9.0  427070339_1413275162   
2  45105.3           6.0            9.0  427070339_1413275162   
3  45105.3           6.0            9.0  427070339_1413275162   
4  45105.3           6.0            9.0  427070339_1413275162   

                      visit_id process_step            date_time Variation  
0  228976764_46825473280_96584      confirm  2017-04-02 11:51:13      Test  
1  228976764_46825473280_96584      confirm  2017-04-02 11:47:50      Test  
2  228976764_46825473280_96584      confirm  2017-04-02 11:46:45      Test  
3  228976764_46825473280_96584       step_3  2017-04-02 11:23:08      Test  
4  228976764_46825473280_96584       step_2  2017-04-02 11:22:24      Test

---
# *EXPLORE & CLEAN vanguard_ab_test dataset:* 

In [11]:
import pandas as pd
df_vanguard_ab_test = pd.read_csv("../data/interim/vanguard_ab_test.csv")

# display vanguard_ab_test dataset
print("Vanguard A/B Test Dataset:")
print("Vanguard A/B Test Dataset Shape:", df_vanguard_ab_test.shape[0], "rows and", df_vanguard_ab_test.shape[1], "columns\n")
df_vanguard_ab_test.info()
display(df_vanguard_ab_test.head())

C:\Users\sboub\AppData\Local\Temp\ipykernel_7220\1209327210.py:2: DtypeWarning: Columns (0: Variation) have mixed types. Specify dtype option on import or set low_memory=False.
  df_vanguard_ab_test = pd.read_csv("../data/interim/vanguard_ab_test.csv")


Vanguard A/B Test Dataset:
Vanguard A/B Test Dataset Shape: 449831 rows and 14 columns

<class 'pandas.DataFrame'>
RangeIndex: 449831 entries, 0 to 449830
Data columns (total 14 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   client_id         449831 non-null  int64  
 1   clnt_tenure_yr    449716 non-null  float64
 2   clnt_tenure_mnth  449716 non-null  float64
 3   clnt_age          449704 non-null  float64
 4   gendr             449716 non-null  str    
 5   num_accts         449716 non-null  float64
 6   bal               449716 non-null  float64
 7   calls_6_mnth      449716 non-null  float64
 8   logons_6_mnth     449716 non-null  float64
 9   visitor_id        449831 non-null  str    
 10  visit_id          449831 non-null  str    
 11  process_step      449831 non-null  str    
 12  date_time         449831 non-null  str    
 13  Variation         321309 non-null  str    
dtypes: float64(7), int64(1), str(6)
memory 

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0     836976             6.0              73.0      60.5     U        2.0   
1     836976             6.0              73.0      60.5     U        2.0   
2     836976             6.0              73.0      60.5     U        2.0   
3     836976             6.0              73.0      60.5     U        2.0   
4     836976             6.0              73.0      60.5     U        2.0   

       bal  calls_6_mnth  logons_6_mnth            visitor_id  \
0  45105.3           6.0            9.0  427070339_1413275162   
1  45105.3           6.0            9.0  427070339_1413275162   
2  45105.3           6.0            9.0  427070339_1413275162   
3  45105.3           6.0            9.0  427070339_1413275162   
4  45105.3           6.0            9.0  427070339_1413275162   

                      visit_id process_step            date_time Variation  
0  228976764_46825473280_96584      confirm  2017-04-02 11:51:13      Test  
1  228976764_46825473280_96584      confirm  2017-04-02 11:47:50      Test  
2  228976764_46825473280_96584      confirm  2017-04-02 11:46:45      Test  
3  228976764_46825473280_96584       step_3  2017-04-02 11:23:08      Test  
4  228976764_46825473280_96584       step_2  2017-04-02 11:22:24      Test

In [13]:
df_vanguard_ab_test.columns.tolist()

['client_id',
 'clnt_tenure_yr',
 'clnt_tenure_mnth',
 'clnt_age',
 'gendr',
 'num_accts',
 'bal',
 'calls_6_mnth',
 'logons_6_mnth',
 'visitor_id',
 'visit_id',
 'process_step',
 'date_time',
 'Variation']

---
## *Standardize columns names:*

| Original Column Name    | Standardized Column Name       |
|-------------------------|-------------------------------|
| `clnt_tenure_yr`        | `client_tenure_years`        |
| `clnt_tenure_mnth`      | `client_tenure_months`       |
| `clnt_age`              | `client_age`                 |
| `gendr`                 | `gender`                     |
| `num_accts`             | `number_of_accounts`         |
| `bal`                   | `account_balance`            |
| `calls_6_mnth`          | `calls_last_6_months`        |
| `logons_6_mnth`         | `logins_last_6_months`       |
| `Variation`             | `test_group`                 |


In [20]:
import pandas as pd

# Load your main Vanguard A/B test dataset
df_vanguard_ab_test = pd.read_csv("../data/interim/vanguard_ab_test.csv")

# Define the renaming dictionary
rename_columns = {
    "clnt_tenure_yr": "client_tenure_years",
    "clnt_tenure_mnth": "client_tenure_months",
    "clnt_age": "client_age",
    "gendr": "gender",
    "num_accts": "number_of_accounts",
    "bal": "account_balance",
    "calls_6_mnth": "calls_last_6_months",
    "logons_6_mnth": "logins_last_6_months",
    "Variation": "test_group"
}

# Rename columns
df_vanguard_ab_test.rename(columns=rename_columns, inplace=True)

# Save the updated dataset
df_vanguard_ab_test.to_csv("../data/interim/vanguard_ab_test_standardized.csv", index=False)


C:\Users\sboub\AppData\Local\Temp\ipykernel_7220\4007725693.py:4: DtypeWarning: Columns (0: Variation) have mixed types. Specify dtype option on import or set low_memory=False.
  df_vanguard_ab_test = pd.read_csv("../data/interim/vanguard_ab_test.csv")


In [21]:
display(df_vanguard_ab_test.head())

client_id  client_tenure_years  client_tenure_months  client_age gender  \
0     836976                  6.0                  73.0        60.5      U   
1     836976                  6.0                  73.0        60.5      U   
2     836976                  6.0                  73.0        60.5      U   
3     836976                  6.0                  73.0        60.5      U   
4     836976                  6.0                  73.0        60.5      U   

   number_of_accounts  account_balance  calls_last_6_months  \
0                 2.0          45105.3                  6.0   
1                 2.0          45105.3                  6.0   
2                 2.0          45105.3                  6.0   
3                 2.0          45105.3                  6.0   
4                 2.0          45105.3                  6.0   

   logins_last_6_months            visitor_id                     visit_id  \
0                   9.0  427070339_1413275162  228976764_46825473280_96584   
1                   9.0  427070339_1413275162  228976764_46825473280_96584   
2                   9.0  427070339_1413275162  228976764_46825473280_96584   
3                   9.0  427070339_1413275162  228976764_46825473280_96584   
4                   9.0  427070339_1413275162  228976764_46825473280_96584   

  process_step            date_time test_group  
0      confirm  2017-04-02 11:51:13       Test  
1      confirm  2017-04-02 11:47:50       Test  
2      confirm  2017-04-02 11:46:45       Test  
3       step_3  2017-04-02 11:23:08       Test  
4       step_2  2017-04-02 11:22:24       Test

In [70]:
df_vanguard_ab_test.info()

<class 'pandas.DataFrame'>
RangeIndex: 449831 entries, 0 to 449830
Data columns (total 14 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   client_id             449831 non-null  int64  
 1   client_tenure_years   449716 non-null  float64
 2   client_tenure_months  449716 non-null  float64
 3   client_age            449704 non-null  float64
 4   gender                449716 non-null  str    
 5   number_of_accounts    449716 non-null  float64
 6   account_balance       449716 non-null  float64
 7   calls_last_6_months   449716 non-null  float64
 8   logins_last_6_months  449831 non-null  int64  
 9   visitor_id            449831 non-null  str    
 10  visit_id              449831 non-null  str    
 11  process_step          449831 non-null  str    
 12  date_time             449831 non-null  str    
 13  test_group            321309 non-null  str    
dtypes: float64(6), int64(2), str(6)
memory usage: 48.0 MB


In [73]:
df_vanguard_ab_test["logins_last_6_months"].dtype

dtype('int64')

#### change columns type:
1.   clnt_tenure_yr    449716 non-null  float64 to int
2.   clnt_tenure_mnth  449716 non-null  float64 to int
3.   clnt_age          449704 non-null  float64 to in
5.   number_of_accounts    449716 non-null  float64 to int
7.   calls_last_6_months   449716 non-null  float64



In [74]:
display(df_vanguard_ab_test.head())

client_id  client_tenure_years  client_tenure_months  client_age gender  \
0     836976                  6.0                  73.0        60.5      U   
1     836976                  6.0                  73.0        60.5      U   
2     836976                  6.0                  73.0        60.5      U   
3     836976                  6.0                  73.0        60.5      U   
4     836976                  6.0                  73.0        60.5      U   

   number_of_accounts  account_balance  calls_last_6_months  \
0                 2.0          45105.3                  6.0   
1                 2.0          45105.3                  6.0   
2                 2.0          45105.3                  6.0   
3                 2.0          45105.3                  6.0   
4                 2.0          45105.3                  6.0   

   logins_last_6_months            visitor_id                     visit_id  \
0                     9  427070339_1413275162  228976764_46825473280_96584   
1                     9  427070339_1413275162  228976764_46825473280_96584   
2                     9  427070339_1413275162  228976764_46825473280_96584   
3                     9  427070339_1413275162  228976764_46825473280_96584   
4                     9  427070339_1413275162  228976764_46825473280_96584   

  process_step            date_time test_group  
0      confirm  2017-04-02 11:51:13       Test  
1      confirm  2017-04-02 11:47:50       Test  
2      confirm  2017-04-02 11:46:45       Test  
3       step_3  2017-04-02 11:23:08       Test  
4       step_2  2017-04-02 11:22:24       Test

---
## *CLEAN DATA*
- df["column_name"].unique()
- data mapping
- replace()


In [46]:
df_vanguard_ab_test.columns.tolist()

['client_id',
 'client_tenure_years',
 'client_tenure_months',
 'client_age',
 'gender',
 'number_of_accounts',
 'account_balance',
 'calls_last_6_months',
 'logins_last_6_months',
 'visitor_id',
 'visit_id',
 'process_step',
 'date_time',
 'test_group']

---
### *1. Checking for unique values in every column*

In [ ]:
# client_id ------------------------------------------------------------------------------
print(f"Client IDs: {df_vanguard_ab_test['client_id'].unique()}\n")

# client_tenure_years ------------------------------------------------------------------------------
print(f"Client Tenure Years: {df_vanguard_ab_test['client_tenure_years'].unique()}\n")

# client_tenure_months ------------------------------------------------------------------------------
print(f"Client Tenure Months: {df_vanguard_ab_test['client_tenure_months'].unique()}\n")

# client_age ------------------------------------------------------------------------------
print(f"Client Age: {df_vanguard_ab_test['client_age'].unique()}\n")

# gender ------------------------------------------------------------------------------
print(f"Gender: {df_vanguard_ab_test['gender'].unique()}\n")
# F is Female
# M is Male
# U is Unknown
# X is binary
# fill nulls with 'U' for Unknown

# number_of_accounts ------------------------------------------------------------------------------
print(f"Number of Accounts: {df_vanguard_ab_test['number_of_accounts'].unique()}\n")

# account_balance ------------------------------------------------------------------------------
print(f"Account Balance: {df_vanguard_ab_test['account_balance'].unique()}\n")

# calls_last_6_months ------------------------------------------------------------------------------
print(f"Calls Last 6 Months: {df_vanguard_ab_test['calls_last_6_months'].unique()}\n")

# logins_last_6_months ------------------------------------------------------------------------------
print(f"Logins Last 6 Months: {df_vanguard_ab_test['logins_last_6_months'].unique()}\n")

# visitor_id ------------------------------------------------------------------------------
print(f"Visitor IDs: {df_vanguard_ab_test['visitor_id'].unique()}\n")

# visit_id ------------------------------------------------------------------------------
print(f"Visit IDs: {df_vanguard_ab_test['visit_id'].unique()}\n")

# session_id ------------------------------------------------------------------------------
print(f"Process Steps: {df_vanguard_ab_test['process_step'].unique()}\n")

# date_time ------------------------------------------------------------------------------
print(f"Date and Time: {df_vanguard_ab_test['date_time'].unique()}\n")

# test_group ------------------------------------------------------------------------------
print(f"Test Groups: {df_vanguard_ab_test['test_group'].unique()}\n")

# 20.5 years of experience
# 20 years + 0.5*12 = 20 years and 6 months
# clean client_age removing comma just decimal number
# 20 years instead of 20.5 years

Client IDs: [ 836976 2304905 1439522 ...  333913 1573142 5602139]

Client Tenure Years: [ 6.  7.  5. 16. 12. 30. 15.  8.  4.  3. 10. 11. 13. 14. 19. 18. 25. 23.
 21. 17.  9. 27. 24. 20. 28. 22. 29. 46. 26. 32. 33. 35.  2. 38. 31. 48.
 34. 36. 51. nan 47. 49. 37. 39. 40. 55. 42. 44. 43. 50. 54. 41. 45. 52.
 62.]

Client Tenure Months: [ 73.  94.  64. 198. 145.  71.  66. 361. 369. 189. 100.  99.  67. 193.
  58.  47. 122. 136. 101. 103.  70. 152. 160.  80.  88.  86.  65.  68.
  93.  85. 168. 142. 228. 223. 305. 286. 285. 280. 254. 253. 278. 221.
 216. 212. 204. 260. 116. 130. 177. 159. 190.  89.  91. 129. 329.  92.
  82. 169. 239.  69. 178.  95.  63. 187. 288. 261. 251. 250.  56.  60.
 123. 175. 188.  59. 336. 149. 156. 330. 155. 167.  96. 242. 163.  81.
 182. 134. 113. 172. 120. 105. 271. 244. 200. 157.  78. 151. 121. 184.
 335. 357. 111. 185.  76. 148.  97.  83. 347. 324. 162. 194.  74. 186.
 300. 289. 295. 282. 270. 246. 229. 206. 109. 343. 176. 144. 161. 179.
 106. 154. 119.  75. 166.

---
# 2.1 Data Quality Issues – Vanguard A/B Test Dataset

This section highlights columns that contain **invalid, inconsistent, or semantically incorrect values**, not just missing data.

---

### 1. `client_tenure_years`

**Issues**

* Contains `NaN` values
*
* incorrect data type float (e.g. 10, 30.5 , 55, 62 years)

**To check**
* Long tenure values may be unrealistic
* Should be consistent with `client_tenure_months`


---

### 2. `client_tenure_months` 🚩

**Issues**

* Contains `NaN` values
* (e.g. 554, 576, 612, 669, 749 months)
* Contains `NaN` values

**Why this is an issue**

* Should be consistent with `client_tenure_years`

**Action Needed**

* Decide whether:

  * Months is about which month `1<=month<= 12`or the number of months
  * `client_tenure_months` should equal `client_tenure_years × 12`
    * Correct or drop inconsistent records t

---

### 3. `client_age` 🚩

**Issues**

* Decimal values (e.g. 32.5, 60.5)
* Very low ages (e.g. 13.5, 14, 15) under age clients
// parents saving accounts for their children
* Very high ages (e.g. 94, 96) elder age clients

**Why this is an issue**

* Financial clients are typically adults (18+)
* Age should be an integer
* Extreme ages may be outliers or data errors

**Action Needed**

* Remove or cap unrealistic ages
* Convert to integer after cleaning

---

### 4. `gender`

**Issues**

* Values include: `M`, `F`, `U`, `X`

**Why this is an issue**

* `U` and `X` are ambiguous and need definition

**Action Needed**

* Normalize values:

  * `M` → Male
  * `F` → Female
  * `U` → Unknown
  * `X` → Other
// check vanguard page to see if there are the same options
---

### 5. `number_of_accounts`

**Issues**

* Stored as float values (e.g. `2.0`, `5.0`)
* Contains `NaN` values

**Why this is an issue**

* Account counts should be non-negative integers

**Action Needed**

* Fill missing values if justified
* Convert to integer type

---

### 6. `account_balance`

**Issues**

* No obvious issues observed

**Notes**

* Should be checked for negative balances depending on business rules

---

### 7. `calls_last_6_months`

**Issues**

* Stored as floats
* Contains `NaN` values

**Why this is an issue**

* Call counts should be integers

**Action Needed**

* Replace `NaN` with `0` if “no calls” is a valid assumption
* Convert to integer type

---

### 8. `logins_last_6_months`

**Issues**

* Stored as floats
* Contains `NaN` values

**Why this is an issue**

* Login counts should be integers

**Action Needed**

* Replace `NaN` with `0` if “no logins” is a valid assumption
* Convert to integer type

---

### 9. `visitor_id`

**Issues**

* None detected

**Status**

* Valid identifier column

---

### 10. `visit_id`

**Issues**

* None detected

**Status**

* Valid identifier column

---

### 11. `process_step`

**Issues**

* Categorical but ordered (`start` → `confirm`)

**Why this is an issue**

* Ordering matters for funnel analysis

**Action Needed**

* Convert to an ordered categorical variable

---

### 12. `date_time`

**Issues**

* Stored as string

**Why this is an issue**

* Prevents time-based analysis

**Action Needed**

* Convert to `datetime` type

---

### 13. `test_group`

**Issues**

* Values: `Test`, `Control`, `NaN`

**Why this is an issue**

* A/B test participants must belong to a group

**Action Needed**

* Remove or reassign rows with missing group values

---

## Summary of Key Issues

| Column Name            | Issue Type                     |
| ---------------------- | ------------------------------ |
| `client_tenure_years`  | Potentially unrealistic values |
| `client_tenure_months` | Logical inconsistency          |
| `client_age`           | Invalid ages and decimals      |
| `gender`               | Ambiguous categories           |
| `number_of_accounts`   | Incorrect data type            |
| `calls_last_6_months`  | Incorrect data type            |
| `logins_last_6_months` | Incorrect data type            |
| `process_step`         | Needs ordering                 |
| `date_time`            | Incorrect data type            |
| `test_group`           | Invalid A/B assignment         |

---

If you want, I can next:

* Write **cleaning code per column**
* Create a **data cleaning pipeline**
* Or help you justify assumptions (e.g. replacing nulls with 0) for your report


---
### *2. Clean invalid values*
Columns with invalid values

---
#### gender
| Data field | invalid value                       | How to correct the invalid values   |
| ----- | ------------------------------------ | ------------------------------------ |
| `gender`   |    nulls                             | fill nulls with U for Unknown    |
| ``   |                                |    |
| ``   |  |    |
| ``   |   |    |
| `` |               |    |


In [69]:
display(df_vanguard_ab_test.info())

<class 'pandas.DataFrame'>
RangeIndex: 449831 entries, 0 to 449830
Data columns (total 14 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   client_id             449831 non-null  int64  
 1   client_tenure_years   449716 non-null  float64
 2   client_tenure_months  449716 non-null  float64
 3   client_age            449704 non-null  float64
 4   gender                449716 non-null  str    
 5   number_of_accounts    449716 non-null  float64
 6   account_balance       449716 non-null  float64
 7   calls_last_6_months   449716 non-null  float64
 8   logins_last_6_months  449831 non-null  int64  
 9   visitor_id            449831 non-null  str    
 10  visit_id              449831 non-null  str    
 11  process_step          449831 non-null  str    
 12  date_time             449831 non-null  str    
 13  test_group            321309 non-null  str    
dtypes: float64(6), int64(2), str(6)
memory usage: 48.0 MB


None

---
#### gender
| Value | Expression                       |
| ----- | ------------------------------------ |
| `M`   | Male                                 |
| `F`   | Female                               |
| `U`   | Unknown / Unspecified / Not provided |
| `X`   | Non-binary / Other / Not classified  |
| `nan` | Missing value (no data) to fill with Unknown              |

In [ ]:
df_vanguard_ab_test["gender"].unique()
# F is Female
# M is Male
# U is Unknown
# X is binary
# fill nulls with 'U' for Unknown
df_vanguard_ab_test['gender'].fillna('U', inplace=True)

C:\Users\sboub\AppData\Local\Temp\ipykernel_7220\2323260372.py:6: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df_vanguard_ab_test['gender'].fillna('U', inplace=True)


0         U
1         U
2         U
3         U
4         U
         ..
449826    F
449827    F
449828    F
449829    F
449830    F
Name: gender, Length: 449831, dtype: str

---
#### logins_last_6_months

In [ ]:
df_vanguard_ab_test["logins_last_6_months"].unique()
# fill nulls with 0 because there is no data of logins so we assume it's no logins
# convert type to int

# Fill NaN with 0 first
df_vanguard_ab_test['logins_last_6_months'] = df_vanguard_ab_test['logins_last_6_months'].fillna(0)

# Convert to int type
df_vanguard_ab_test['logins_last_6_months'] = df_vanguard_ab_test['logins_last_6_months'].astype(int)

array([9, 6, 3, 5, 4, 8, 7, 0, 1, 2])

In [67]:
# Check unique values
print(df_vanguard_ab_test["logins_last_6_months"].unique())


[9 6 3 5 4 8 7 0 1 2]


In [58]:
# Fill NaN with 0 first
df_vanguard_ab_test['logins_last_6_months'] = df_vanguard_ab_test['logins_last_6_months'].fillna(0)

# Now convert to int
df_vanguard_ab_test['logins_last_6_months'] = df_vanguard_ab_test['logins_last_6_months'].astype(int)


In [60]:
print(df_vanguard_ab_test["logins_last_6_months"])

0         9
1         9
2         9
3         9
4         9
         ..
449826    7
449827    7
449828    7
449829    7
449830    7
Name: logins_last_6_months, Length: 449831, dtype: int64


In [65]:
display(df_vanguard_ab_test.head())

client_id  client_tenure_years  client_tenure_months  client_age gender  \
0     836976                  6.0                  73.0        60.5      U   
1     836976                  6.0                  73.0        60.5      U   
2     836976                  6.0                  73.0        60.5      U   
3     836976                  6.0                  73.0        60.5      U   
4     836976                  6.0                  73.0        60.5      U   

   number_of_accounts  account_balance  calls_last_6_months  \
0                 2.0          45105.3                  6.0   
1                 2.0          45105.3                  6.0   
2                 2.0          45105.3                  6.0   
3                 2.0          45105.3                  6.0   
4                 2.0          45105.3                  6.0   

   logins_last_6_months            visitor_id                     visit_id  \
0                     9  427070339_1413275162  228976764_46825473280_96584   
1                     9  427070339_1413275162  228976764_46825473280_96584   
2                     9  427070339_1413275162  228976764_46825473280_96584   
3                     9  427070339_1413275162  228976764_46825473280_96584   
4                     9  427070339_1413275162  228976764_46825473280_96584   

  process_step            date_time test_group  
0      confirm  2017-04-02 11:51:13       Test  
1      confirm  2017-04-02 11:47:50       Test  
2      confirm  2017-04-02 11:46:45       Test  
3       step_3  2017-04-02 11:23:08       Test  
4       step_2  2017-04-02 11:22:24       Test

In [55]:

df_vanguard_ab_test['logins_last_6_months'].dtype

dtype('float64')

In [51]:
print(df_vanguard_ab_test["logins_last_6_months"].head(20))

0     9.0
1     9.0
2     9.0
3     9.0
4     9.0
5     9.0
6     9.0
7     9.0
8     9.0
9     9.0
10    9.0
11    9.0
12    9.0
13    9.0
14    9.0
15    9.0
16    9.0
17    9.0
18    9.0
19    9.0
Name: logins_last_6_months, dtype: float64


---
#### visitor_id

In [ ]:
df_vanguard_ab_test["visitor_id"].unique()

<StringArray>
[ '427070339_1413275162', '336210275_56978226960', '245908175_84333642537',
 '407067363_48226131468', '391202328_71710993352', '432077852_35036160885',
 '875448077_54387937735', '227827901_46893415084',  '56607555_75044897582',
 '667009768_32934260678',
 ...
 '616643929_99035070625',  '87972764_45157109967',  '43214017_97463037638',
  '55178744_59935472472', '441324114_37855673027',  '235892744_2995807940',
 '405457094_54883289172', '110929964_38003873994',  '880662048_9504251369',
 '962766819_57063135389']
Length: 78187, dtype: str

---
#### visit_id

In [35]:
df_vanguard_ab_test["visit_id"].unique()

<StringArray>
[ '228976764_46825473280_96584',  '104438405_2368283624_817211',
 '724930510_65323235593_692754', '898099666_87299892721_263146',
  '968288276_7230824336_666089',  '71954691_44810191073_495385',
 '942996157_39527225999_589383', '717519432_65547872352_613051',
 '468902637_55778357932_556682', '247503697_11679931698_453228',
 ...
 '758008859_31261979266_426270', '458308265_80160772759_795974',
  '10432687_78618454141_814606', '504342110_43310803922_587933',
 '535967473_99843170075_417938', '709721715_35968917131_979675',
  '326174177_47892407008_51741', '448875043_84914821939_263970',
   '158544612_4105226940_43043', '884309593_49820690623_654151']
Length: 96773, dtype: str

---
#### process_step

In [24]:
df_vanguard_ab_test["process_step"].unique()

<StringArray>
['confirm', 'step_3', 'step_2', 'step_1', 'start']
Length: 5, dtype: str

---
#### date_time

In [ ]:
df_vanguard_ab_test["date_time"].unique()

<StringArray>
['2017-04-02 11:51:13', '2017-04-02 11:47:50', '2017-04-02 11:46:45',
 '2017-04-02 11:23:08', '2017-04-02 11:22:24', '2017-04-02 11:21:38',
 '2017-04-02 11:21:28', '2017-03-29 11:02:44', '2017-03-29 11:01:40',
 '2017-03-29 10:59:43',
 ...
 '2017-05-31 06:13:59', '2017-04-09 13:03:40', '2017-04-09 12:46:22',
 '2017-04-09 12:44:25', '2017-04-06 14:22:03', '2017-04-06 14:19:28',
 '2017-04-19 12:11:16', '2017-04-24 11:45:37', '2017-04-24 11:44:51',
 '2017-04-24 11:44:36']
Length: 384038, dtype: str

---
### *3. Deal with Nulls (Standardized Dataset)*
Columns with Null Values 
| Column Number | Column Name             |
|---------------|------------------------|
| 1             | `client_tenure_years`  |
| 2             | `client_tenure_months` |
| 3             | `client_age`           |
| 4             | `gender`               |
| 5             | `number_of_accounts`   |
| 6             | `account_balance`      |
| 7             | `calls_last_6_months`  |
| 8             | `logins_last_6_months` |
| 13            | `test_group`           |


---
#### gender
fill nulls with "U"

| Value | Likely Meaning                       |
| ----- | ------------------------------------ |
| `M`   | Male                                 |
| `F`   | Female                               |
| `U`   | Unknown / Unspecified / Not provided |
| `X`   | Non-binary / Other / Not classified  |
| `nan` | Missing value (no data) to fill with Unknown              |

In [ ]:
df_vanguard_ab_test["gender"].unique()
# F is Female
# M is Male
# U is Unknown
# X is binary
# fill nulls with 'U' for Unknown
df_vanguard_ab_test['gender'].fillna('U', inplace=True)

C:\Users\sboub\AppData\Local\Temp\ipykernel_7220\2323260372.py:6: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df_vanguard_ab_test['gender'].fillna('U', inplace=True)


0         U
1         U
2         U
3         U
4         U
         ..
449826    F
449827    F
449828    F
449829    F
449830    F
Name: gender, Length: 449831, dtype: str

---
#### logins_last_6_months

In [ ]:
df_vanguard_ab_test["logins_last_6_months"].unique()
# fill nulls with 0 because there is no data of logins so we assume it's no logins
# convert type to int

array([9, 6, 3, 5, 4, 8, 7, 0, 1, 2])

---
#### test_group

In [ ]:
df_vanguard_ab_test["test_group"].unique()
# deal with nulls that don't fit in any group

<StringArray>
['Test', 'Control', nan]
Length: 3, dtype: str

---
## 4. Deal with duplicates

---
---
# A/B ANALYSIS & PLOTS

In [66]:
display(df_vanguard_ab_test.describe())

client_id  client_tenure_years  client_tenure_months     client_age  \
count  4.498310e+05        449716.000000         449716.000000  449704.000000   
mean   5.002870e+06            12.234130            152.810921      47.653966   
std    2.873031e+06             7.005183             83.719616      15.742089   
min    1.690000e+02             2.000000             33.000000      13.500000   
25%    2.515550e+06             6.000000             82.000000      33.500000   
50%    5.025522e+06            11.000000            139.000000      49.000000   
75%    7.468377e+06            16.000000            194.000000      60.000000   
max    9.999839e+06            62.000000            749.000000      96.000000   

       number_of_accounts  account_balance  calls_last_6_months  \
count       449716.000000     4.497160e+05        449716.000000   
mean             2.264767     1.597737e+05             3.529176   
std              0.543074     3.420694e+05             2.242472   
min              1.000000     1.378942e+04             0.000000   
25%              2.000000     3.844801e+04             2.000000   
50%              2.000000     6.677490e+04             3.000000   
75%              2.000000     1.490039e+05             6.000000   
max              8.000000     1.632004e+07             7.000000   

       logins_last_6_months  
count         449831.000000  
mean               5.708417  
std                2.357644  
min                0.000000  
25%                4.000000  
50%                6.000000  
75%                8.000000  
max                9.000000